src/json.ts

```diff
@@ -34,6 +40,109 @@ export type Json =
   | Json[]
   | { [prop: string]: Json };
 
+/**
+ * A helper type to make properties with `undefined` in their type optional, but
+ * not `undefined` itself.
+ *
+ * @example
+ * ```ts
+ * type Foo = ObjectOptional<{ foo: string | undefined }>;
+ * // Foo is equivalent to { foo?: string }
+ * ```
+ */
+export type ObjectOptional<Schema extends Record<string, unknown>> = {
+  [Key in keyof Schema as Schema[Key] extends ExactOptionalGuard
+    ? Key
+    : never]?: Schema[Key] extends ExactOptionalGuard & infer Original
```

Comment:
IMO, if `exactOptionalPropertyTypes` is disabled, the type should be left as

```typescript
{ foo: string }
```

instead of

```typescript
{ foo?: string | undefined }
```

Reply:
I'm not 100% sure about it. It would make it more difficult to work with if `exactOptionalPropertyTypes` is disabled (which it is currently on the Snaps repo).